src/EditorFeatures/Core/Implementation/Suggestions/SuggestedAction.cs

```diff
@@ -102,44 +102,45 @@ public void Invoke(CancellationToken cancellationToken)
 
         private async Task YieldThenInvokeAsync(CancellationToken cancellationToken)
         {
-            this.AssertIsForeground();
+            AssertIsForeground();
 
             // Always wrap whatever we're doing in a threaded wait dialog.
-            using (var context = this.SourceProvider.WaitIndicator.StartWait(CodeAction.Title, CodeAction.Message, allowCancel: true, showProgress: true))
-            using (var linkedSource = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken, context.CancellationToken))
+            using (var waitContext = await GetWaitContextAsync(cancellationToken).ConfigureAwait(true))
+            using (var linkedSource = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken, waitContext.CancellationToken))
             {
-                // Yield the UI thread so that the light bulb can be dismissed.  This is necessary
-                // as some code actions may be long running, and we don't want the light bulb to
-                // stay on screen.
-                await Task.Yield();
-
-                this.AssertIsForeground();
+                AssertIsForeground();
 
                 // Then proceed and actually do the invoke.
-                await InvokeAsync(context.ProgressTracker, linkedSource.Token).ConfigureAwait(true);
+                InvokeAsync(waitContext.ProgressTracker, linkedSource.Token).Wait(linkedSource.Token);
```

Comment:
instead of doing the blocking wait here, can we instead move the 'Wait' up to 'Invoke'?  That way 'Invoke' stays synchronous and prevents users from interacting with the UI.  On the plus side, we also won't have any blocking in these async methods.

Note: we'll need a try/catch since Wait(...) can then throw cancellation exceptions.